File: 0120-triangle/0120-triangle.cpp
class Solution {
public:
   vector<vector<int>>dp;
    int solve(vector<vector<int>>& triangle,int m , int n){
        int last =triangle.size();
        if(m==last-1)return triangle[m][n];
        if(dp[m][n]!=INT_MAX)return dp[m][n];

        int down = triangle[m][n] + solve(triangle,m+1,n);
        int dia = triangle[m][n] + solve(triangle,m+1,n+1);

        return dp[m][n]=min(dia,down);
    }
    int minimumTotal(vector<vector<int>>& triangle) {
        int m = triangle.size();
        int n = triangle[m-1].size();
        dp.assign(m,vector<int>(n,INT_MAX));
        return  solve(triangle,0,0);
    }
};

File: 0120-triangle/0120-triangle.py
class Solution:
    dp=[]
    def solve(self,triangle,row,col):
        m=len(triangle)
        n=len(triangle[m-1])
        if(row==m-1):
            return triangle[row][col]
        if self.dp[row][col] != float("inf") :
            return self.dp[row][col]
        down=triangle[row][col]+self.solve(triangle,row+1,col)
        dia=triangle[row][col]+self.solve(triangle,row+1,col+1)

        self.dp[row][col]=min(down,dia)
        return self.dp[row][col]
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        m=len(triangle)
        n=len(triangle[m-1])
        self.dp=[[float("inf")]*n for i in range(0,m)]

        return self.solve(triangle,0,0)